File: exercism/haskell/crypto-square/CryptoSquare.hs
module CryptoSquare (normalizePlaintext,
                     squareSize,
                     plaintextSegments,
                     ciphertext,
                     normalizeCiphertext) where

import Data.Char (toLower, isAlphaNum)
import Data.List (transpose)
import Data.List.Split (chunksOf)

normalizePlaintext :: String -> String
normalizePlaintext = map toLower . filter isAlphaNum

squareSize :: String -> Int
squareSize x = ceiling $ sqrt len
  where len = fromIntegral (length x) :: Float

plaintextSegments :: String -> [String]
plaintextSegments text = chunksOf (squareSize cleaned) cleaned
  where cleaned = normalizePlaintext text

ciphertext :: String -> String
ciphertext = concat . transpose . plaintextSegments

normalizeCiphertext :: String -> String
normalizeCiphertext = unwords . transpose . plaintextSegments

File: exercism/csharp/crypto-square/CryptoSquare.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Crypto
{
    public string NormalizePlaintext;
    public int Size;

    public Crypto (string given)
    {
        NormalizePlaintext = String.Concat(given.Where(Char.IsLetterOrDigit)).ToLower();
        Size = (int)Math.Ceiling(Math.Sqrt(NormalizePlaintext.Length));
    }

    public IEnumerable<IEnumerable<char>> PlaintextSegments ()
    {
        IEnumerable<char> text = NormalizePlaintext;
        while (text.Any())
        {
            yield return text.Take(Size);
            text = text.Skip(Size);
        }
    }

    public string Ciphertext ()
    {
        var segments = PlaintextSegments();
        return String.Concat(Transposed(segments));
    }

    public string NormalizeCiphertext ()
    {
        var segments = PlaintextSegments();
        return String.Join(" ", Transposed(segments));
    }

    private IEnumerable<string> Transposed (IEnumerable<IEnumerable<char>> segments)
    {
        foreach (var i in Enumerable.Range(0, Size))
        {
            var s = segments
                .Select(xs => xs.ElementAtOrDefault(i))
                .Where(ch => ch != '\0');
            yield return String.Concat(s);
        }
    }
}
